perspective.el

```diff
@@ -968,6 +990,249 @@ perspective beginning with the given letter."
+    ;; recurse
+    (t (cons (car entry) (cl-loop for e in (cdr entry)
+                               collect (persp--state-window-state-massage e persp valid-buffers))))))
+
+(defun persp--state-frame-data ()
+  (cl-loop for frame in (frame-list)
+           if (frame-parameter frame 'persp--hash) ; XXX: filter non-perspective-enabled frames
+           collect (with-selected-frame frame
+                     (lexical-let ((persps-in-frame (make-hash-table :test 'equal))
+                                   (persp-names-in-order (persp-names)))
+                       (cl-loop for persp in persp-names-in-order do
+                                (unless (persp-killed-p (gethash persp (perspectives-hash)))
+                                  (with-perspective persp
+                                    (lexical-let* ((buffers
+                                                    (cl-loop for buffer in (persp-buffers (persp-curr))
+                                                             if (persp--state-interesting-buffer-p buffer)
+                                                             collect (buffer-name buffer)))
+                                                   (windows
+                                                    (cl-loop for entry in (window-state-get (frame-root-window) t)
+                                                             collect (persp--state-window-state-massage entry persp buffers))))
+                                      (puthash persp
+                                               (make-persp--state-single
+                                                :buffers buffers
+                                                :windows windows)
+                                               persps-in-frame)))))
+                       (make-persp--state-frame
+                        :persps persps-in-frame
+                        :order persp-names-in-order)))))
+
+;;;###autoload
+(cl-defun persp-state-save (&optional file interactive?)
+  "Save the current perspective state to FILE.
+
+FILE defaults to the value of persp-state-default-file if it is
+set.
+
+Each perspective's buffer list and window layout will be saved.
+Frames and their associated perspectives will also be saved
+(but not the original frame sizes).
+
+Buffers with * characters in their names, as well as buffers without
+associated files will be ignored. If such buffers are currently
+visible in a perspective as windows, they will be saved as
+'*scratch* (persp)' buffers."
+  (interactive (list
+                (read-file-name "Save perspective state to file: "
+                                persp-state-default-file
+                                persp-state-default-file)
+                t))
+  (unless persp-mode
+    (message "persp-mode not enabled, nothing to save")
+    (return-from persp-state-save))
+  (lexical-let ((target-file (if (and file (not (string-equal "" file)))
```

Comment:
Should we just do a full lexical conversion?

Reply:
Yes, that's what #99 is about. I'd rather do that as a separate step later.